header.go

```diff
@@ -16,17 +17,42 @@ type HeaderProvider func(*http.Request) (headerName string, headerValue string)
 
 // RoundTrip annotates the outgoing request and calls the wrapped Client.
 func (c *Header) RoundTrip(r *http.Request) (*http.Response, error) {
-	var name, value = c.provider(r)
-	r.Header.Set(name, value)
-	return c.wrapped.RoundTrip(r)
+	if c.requestProvider != nil {
+		var name, value = c.requestProvider(r)
+		r.Header.Set(name, value)
+	}
+	resp, err := c.wrapped.RoundTrip(r)
+	if err != nil {
+		return nil, err
+	}
+	if c.responseProvider != nil {
+		var name, value = c.responseProvider(resp)
+		resp.Header.Set(name, value)
+	}
+	return resp, nil
+}
+
+// NewHeader wraps a transport in order to include custom request headers.
+func NewHeader(requestProvider HeaderProvider) func(http.RoundTripper) http.RoundTripper {
+	return func(c http.RoundTripper) http.RoundTripper {
+		return &Header{
+			wrapped:         c,
+			requestProvider: requestProvider,
+		}
+	}
 }
 
-// NewHeader wraps a transport in order to include custom headers.
-func NewHeader(provider HeaderProvider) func(http.RoundTripper) http.RoundTripper {
+// ResponseHeaderProvider is mapping function that generates the required header name
+// and value to an outgoing response.
+type ResponseHeaderProvider func(*http.Response) (headerName string, headerValue string)
```

Comment:
suggestion: It may be advantageous to accept both the request and the response here for cases where you need to select headers based on the transaction that occurred.